### PR TITLE
S2326, S3925: Add support for record struct

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**

### CODE REVIEW

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
Just some minor suggestions and a few general questions about some coding patterns.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/GenericTypeParameterUnused.cs [hunk 1]
```diff
@@ -37,123 +37,82 @@ public sealed class GenericTypeParameterUnused : SonarDiagnosticAnalyzer
...
+                 SyntaxKindEx.RecordStructDeclaration,
+                 SyntaxKind.StructDeclaration);
+        }
 
         private static void CheckGenericTypeParameters(ISymbol symbol, SyntaxNodeAnalysisContext c)
         {
-            var helper = CreateParametersInfo(c.Node, c.SemanticModel);
-            if (helper.Parameters == null || helper.Parameters.Parameters.Count == 0)
+            var info = CreateParametersInfo(c.Node, c.SemanticModel);
+            if (info?.Parameters is null || info.Parameters.Parameters.Count == 0)
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/GenericTypeParameterUnused.cs:77]: That is not the same. We want to return also for `info is null` and `info.Parametrs is null`

File: analyzers/src/SonarAnalyzer.CSharp/Rules/GenericTypeParameterUnused.cs [hunk 2]
```diff
@@ -37,123 +37,82 @@ public sealed class GenericTypeParameterUnused : SonarDiagnosticAnalyzer
...
-            return methodSymbol != null && methodSymbol.IsChangeable();
-        }
+        private static bool IsMethodCandidate(MethodDeclarationSyntax methodDeclaration, SemanticModel semanticModel) =>
+            !methodDeclaration.Modifiers.Any(x => MethodModifiersToSkip.Contains(x.Kind()))
+            && methodDeclaration.ExplicitInterfaceSpecifier is null
+            && methodDeclaration.HasBodyOrExpressionBody()
+            && semanticModel.GetDeclaredSymbol(methodDeclaration) is { } methodSymbol
+            && methodSymbol.IsChangeable();
 
         private static List<string> GetUsedTypeParameters(IEnumerable<SyntaxNode> declarations, string[] typeParameterNames, SyntaxNodeAnalysisContext context) =>
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/GenericTypeParameterUnused.cs:107]: No, we don't. And I agree with you (and sometimes ask for this change).

In general, I prefer global/nonchanging stuff first. And alternating end last. Especially in table-like situations:
```
Method(c, someArg, "key", true, x => x.Something);
Method(c, someArg, "value", false, x => x.Child.Value.Count());
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/GenericTypeParameterUnused.cs [hunk 3]
```diff
@@ -37,123 +37,82 @@ public sealed class GenericTypeParameterUnused : SonarDiagnosticAnalyzer
...
-
-            var methodSymbol = semanticModel.GetDeclaredSymbol(methodDeclaration);
-
-            return methodSymbol != null && methodSymbol.IsChangeable();
-        }
+        private static bool IsMethodCandidate(MethodDeclarationSyntax methodDeclaration, SemanticModel semanticModel) =>
+            !methodDeclaration.Modifiers.Any(x => MethodModifiersToSkip.Contains(x.Kind()))
+            && methodDeclaration.ExplicitInterfaceSpecifier is null
+            && methodDeclaration.HasBodyOrExpressionBody()
+            && semanticModel.GetDeclaredSymbol(methodDeclaration) is { } methodSymbol
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/GenericTypeParameterUnused.cs:104]: SemanticModel is bound to a SyntaxTree (a file). `symbol.DeclaringSyntaxReferences` can return a node from another SyntaxTree for partial classes. So we need `EnsureCorrectSemanticModelOrDefault` to transition from a known semantic model to the correct one, belonging to that node.

We need it in `GetUsedTypeParameters`, because this method processes result of `symbol.DeclaringSyntaxReferences`.

We don't need it in `IsMethodCandidate`, because this one processes result of `SyntaxNodeAnalysisContext` items `Node` and `SemanticModel`, so it is by-design guaranteed that it is the correct semantic model.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ImplementISerializableCorrectly.cs [hunk 4]
```diff
@@ -111,116 +101,73 @@ private static IEnumerable<SecondaryLocation> CheckGetObjectData(TypeDeclaration
...
 
-            var baseKeyword = constructorDeclaration.Initializer?.ThisOrBaseKeyword;
-            return baseKeyword.HasValue && baseKeyword.Value.IsKind(SyntaxKind.BaseKeyword);
-        }
+        private static bool IsCallingBaseConstructor(IMethodSymbol constructorSymbol) =>
+            (ConstructorDeclarationSyntax)constructorSymbol.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax() is { } constructorDeclaration
+            && constructorDeclaration.Initializer?.ThisOrBaseKeyword is { } baseKeyword
+            && baseKeyword.IsKind(SyntaxKind.BaseKeyword);
 
         private static bool ImplementsISerializable(ITypeSymbol typeSymbol) =>
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ImplementISerializableCorrectly.cs:172]: How can we use `FindImplementationForInterfaceMember`? That one starts from a member symbol and we have only type symbol.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ImplementISerializableCorrectly.cs [hunk 5]
```diff
@@ -53,53 +53,43 @@ public class ImplementISerializableCorrectly : SonarDiagnosticAnalyzer
...
         private static IEnumerable<TSyntax> DeclarationOrImplementation<TSyntax>(TypeDeclarationSyntax typeDeclaration, IMethodSymbol symbol) =>
-            symbol.PartialImplementationPart != null
-                   && symbol.PartialImplementationPart.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax() is { } partialImplementation
-                   && typeDeclaration.DescendantNodes().Any(x => x.Equals(partialImplementation))
+            symbol.PartialImplementationPart is not null
+            && symbol.PartialImplementationPart.DeclaringSyntaxReferences.First().GetSyntax() is var partialImplementation
+            && typeDeclaration.DescendantNodes().Any(x => x.Equals(partialImplementation))
                 ? new[] { partialImplementation }.Cast<TSyntax>()
-                : symbol.DeclaringSyntaxReferences.Select(r => r.GetSyntax()).Cast<TSyntax>();
+                : symbol.DeclaringSyntaxReferences.Select(x => x.GetSyntax()).Cast<TSyntax>();
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ImplementISerializableCorrectly.cs:92]: I tried to extract common logic around `PartialImplementationPart` in #5603.

This part cannot use that, because it does something very specific to this rule. This piece of code is not reusable by any other rule and is quite weird (related to not-reporting on specific secondary locations). So I'll leave this one as is.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ImplementISerializableCorrectly.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ImplementISerializableCorrectly.cs:170]: Maybe like so:
```cs
var iSerializableSymbol = compilation.GetTypeByMetadataName("System.Runtime.Serialization.ISerializable");
var someMember = iSerializableSymbol.GetMembers().First();
typeSymbol.FindImplementationForInterfaceMember(someMember);
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ImplementISerializableCorrectly.CSharp9.Part1.cs [hunk 6]
```diff
@@ -11,6 +11,22 @@ public record SerializableRecord : ISerializable
...
+    [Serializable]
+    public record SerializableRecord_Positional(SerializableRecord PositionalSerializableField) : ISerializable
+    {
+        public SerializableRecord_Positional() : this((SerializableRecord)null) { }
+        protected SerializableRecord_Positional(SerializationInfo info, StreamingContext context) : this((SerializableRecord)null) { }
+        public virtual void GetObjectData(SerializationInfo info, StreamingContext context) { }
+    }
+
+    [Serializable]
+    public record SerializableRecord_PositionalNonserializable(NonSerializedType PositionalNonserializableField) : ISerializable    // FN, should raise similar rule not to put non-serializable parameters here. It cannot be annotated with [NonSerialized]
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ImplementISerializableCorrectly.CSharp9.Part1.cs:23]: I think the comment about the attributes is not valid. See the [spec](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/records#properties) and the sample:

https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQMwAIBOBTAxge02HQGEAKAbQDMBLbAG1HQDl8A7AZW0xoEN6aAL2zAAugFEAHrmwAHAC4126KQEp0Ab2QBfIA=

File: analyzers/src/SonarAnalyzer.CSharp/Rules/GenericTypeParameterUnused.cs [hunk 7]
```diff
@@ -37,123 +38,82 @@ public sealed class GenericTypeParameterUnused : SonarDiagnosticAnalyzer
...
-                     SyntaxKind.ClassDeclaration,
-                     SyntaxKindEx.RecordClassDeclaration);
-            });
+                         CheckGenericTypeParameters(c, c.ContainingSymbol);
+                     }
+                 },
+                 SyntaxKind.ClassDeclaration,
+                 SyntaxKindEx.RecordClassDeclaration,
+                 SyntaxKindEx.RecordStructDeclaration,
+                 SyntaxKind.StructDeclaration);
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/GenericTypeParameterUnused.cs:72]: Does it make sense to add interfaces too? An interface, that declares a type parameter, but it doesn't use it is at least a code smell IMO.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ImplementISerializableCorrectly.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ImplementISerializableCorrectly.cs:170]: The cardinality of the `.First()` part doesn't make sense to me. We're looking for an interface directly now. While the `FindImplementationForInterfaceMember` would look for a consequence (a member) of presence of that interface. While it still might not exist yet in the IDE (not sure how it behaves there)

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ImplementISerializableCorrectly.CSharp10.cs [hunk 8]
```diff
@@ -7,27 +7,33 @@ public class NonSerializedType
     {
     }
 
-    public struct SerializableStruct_NoAttribute : ISerializable // FN
+    public struct SerializableStruct_NoAttribute : ISerializable    // Noncompliant {{Update this implementation of 'ISerializable' to conform to the recommended serialization pattern.}}
+                                                                    // Secondary@-1 {{Add 'System.SerializableAttribute' attribute on 'SerializableStruct_NoAttribute' because it implements 'ISerializable'.}}
+                                                                    // Secondary@-2 {{Add a 'private' constructor 'SerializableStruct_NoAttribute(SerializationInfo, StreamingContext)'.}}
     {
-        private readonly NonSerializedType field = null; // FN, should be marked with [NonSerialized]
+        private readonly NonSerializedType field = null; // Should be marked with [NonSerialized]
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ImplementISerializableCorrectly.CSharp10.cs:14]: Why did you remove the FN? (Same for the other removals below).

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/GenericTypeParameterUnused.cs [hunk 9]
```diff
@@ -180,11 +180,41 @@ public void Method(Tests.Diagnostics.T x)
...
+
+    public interface IUsedAsArgument<T>
+    {
+        object Create(T arg);
+    }
+
+    public interface IUnusedEmpty<T>    // Noncompliant {{'T' is not used in the interface.}}
+    {
+    }
+
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/GenericTypeParameterUnused.cs:213]:
```suggestion
public interface IUnusedWithVariance<out T>
{
}

public interface IUsedInBaseType<T>: IEnumerable<T>
{
}

public interface IUsedInTypeConstraint<T> where T: class
{
}
```

File: analyzers/its/expected/akka.net/Akka.DistributedData--netstandard2.0-S2326.json [hunk 10]
```diff
@@ -0,0 +1,17 @@
+{
+"issues":  [
+{
+"id":  "S2326",
+"message":  "'T' is not used in the interface.",
+"location":  {
+"uri":  "sources\akka.net\src\contrib\cluster\Akka.DistributedData\Key.cs",
```
Comment (martin-strecker-sonarsource) [inline analyzers/its/expected/akka.net/Akka.DistributedData--netstandard2.0-S2326.json:7]: That's a false positive:
```cs
public interface IKey<out T> : IKey where T : IReplicatedData { } // Compliant. Used in a type constraint
```

File: analyzers/its/expected/akka.net/Akka--netstandard2.0-S2326.json [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+"endLine":  18,
+"endColumn":  45
+}
+}
+},
+{
+"id":  "S2326",
+"message":  "'TQueue' is not used in the interface.",
+"location":  {
+"uri":  "sources\akka.net\src\core\Akka\Dispatch\Mailbox.cs",
```
Comment (martin-strecker-sonarsource) [inline analyzers/its/expected/akka.net/Akka--netstandard2.0-S2326.json:20]: False positives. See below.

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ImplementISerializableCorrectly.CSharp10.cs (hunk 8 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ImplementISerializableCorrectly.CSharp10.cs:14]: I was under the wrong impression that the secondary location messages tell to add it.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/GenericTypeParameterUnused.cs (hunk 9 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/GenericTypeParameterUnused.cs:213]: As discussed offline, I've added them as noncompliant. They are not practically used by the interface.

File: analyzers/its/expected/akka.net/Akka.DistributedData--netstandard2.0-S2326.json (hunk 10 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/its/expected/akka.net/Akka.DistributedData--netstandard2.0-S2326.json:7]: As discussed offline, we'll declare it as TP

File: analyzers/its/expected/akka.net/Akka--netstandard2.0-S2326.json (hunk 11 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/its/expected/akka.net/Akka--netstandard2.0-S2326.json:20]: As discussed offline, we'll declare it as TP

**Review by martin-strecker-sonarsource — APPROVED** [review]
